File: src/store.cpp
// Alpacca — local model store. MIT License. See LICENSE.
#include "store.h"

#include <algorithm>
#include <cctype>
#include <cstring>
#include <ctime>
#include <fstream>
#include <sstream>

namespace alpacca {

// ---- references ---------------------------------------------------------

static bool looks_like_path(const std::string & s) {
    if (starts_with(s, "/") || starts_with(s, "./") || starts_with(s, "../") || starts_with(s, "~/")) {
        return true;
    }
#ifdef _WIN32
    if (s.size() >= 2 && isalpha((unsigned char) s[0]) && s[1] == ':') { // C:\...
        return true;
    }
    if (starts_with(s, ".\\") || starts_with(s, "..\\") || starts_with(s, "\\\\")) {
        return true;
    }
#endif
    if (ends_with(to_lower(s), ".gguf")) {
        std::error_code ec;
        if (fs::exists(s, ec)) {
            return true;
        }
    }
    return false;
}

static fs::path expand_tilde(const std::string & s) {
    if (starts_with(s, "~/")) {
        return fs::path(env_or("HOME", env_or("USERPROFILE", ""))) / s.substr(2);
    }
    return fs::path(s);
}

// Split "name[:tag]" — the tag is whatever follows the last ':' as long as it
// contains no '/', so "hf:org/repo" keeps its prefix handling elsewhere.
static void split_tag(const std::string & in, std::string & name, std::string & tag) {
    size_t colon = in.rfind(':');
    if (colon != std::string::npos && in.find('/', colon) == std::string::npos) {
        name = in.substr(0, colon);
        tag = in.substr(colon + 1);
    } else {
        name = in;
        tag = "";
    }
}

ModelRef parse_model_ref(const std::string & raw) {
    std::string input = trim(raw);
    if (input.empty()) {
        fail("empty model name");
    }

    ModelRef ref;

    if (looks_like_path(input)) {
        ref.source = ModelRef::FILE;
        ref.file = expand_tilde(input);
        return ref;
    }

    bool forced_hf = false;
    bool forced_ollama = false;
    for (const char * p : { "hf:", "hf.co/", "huggingface.co/", "https://huggingface.co/" }) {
        if (starts_with(input, p)) {
            input = input.substr(strlen(p));
            forced_hf = true;
            break;
        }
    }
    if (!forced_hf && starts_with(input, "ollama:")) {
        input = input.substr(7);
        forced_ollama = true;
    }

    std::string name_part, tag;
    split_tag(input, name_part, tag);
    auto parts = split(name_part, '/');
    for (const auto & p : parts) {
        if (trim(p).empty()) {
            fail("invalid model reference: '" + raw + "'");
        }
    }

    if (forced_hf || (!forced_ollama && parts.size() >= 2)) {
        if (parts.size() != 2) {
            fail("Hugging Face references must look like org/repo[:quant] (got '" + raw + "')");
        }
        ref.source = ModelRef::HF;
        ref.ns = parts[0];
        ref.name = parts[1];
        ref.tag = tag; // quant or file selector; empty → default preference
        return ref;
    }

    ref.source = ModelRef::OLLAMA;
    if (parts.size() == 1) {
        ref.ns = "library";
        ref.name = parts[0];
    } else if (parts.size() == 2) {
        ref.ns = parts[0];
        ref.name = parts[1];
    } else {
        fail("Ollama references must look like [user/]name[:tag] (got '" + raw + "')");
    }
    ref.tag = tag.empty() ? "latest" : tag;
    return ref;
}

std::string ModelRef::display() const {
    switch (source) {
        case FILE:
            return file.string();
        case HF: {
            std::string s = "hf:" + ns + "/" + name;
            if (!tag.empty()) {
                s += ":" + tag;
            }
            return s;
        }
        case OLLAMA:
        default: {
            std::string s = (ns == "library") ? name : ns + "/" + name;
            if (tag != "latest") {
                s += ":" + tag;
            }
            return s;
        }
    }
}

fs::path ModelRef::store_dir(const fs::path & root) const {
    switch (source) {
        case HF:
            return root / "hf" / sanitize_component(ns) / sanitize_component(name) /
                   sanitize_component(tag.empty() ? "default" : tag);
        case OLLAMA:
            return root / "ollama" / sanitize_component(ns) / sanitize_component(name) /
                   sanitize_component(tag);
        case FILE:
        default:
            fail("file references have no store directory");
    }
}

// ---- store --------------------------------------------------------------

fs::path models_root() {
    return alpacca_home() / "models";
}

static std::string read_file_text(const fs::path & p) {
    std::ifstream f(p, std::ios::binary);
    if (!f) {
        return "";
    }
    std::ostringstream ss;
    ss << f.rdbuf();
    return ss.str();
}

static bool load_manifest(const fs::path & dir, JValue & out) {
    std::string text = read_file_text(dir / "manifest.json");
    if (text.empty()) {
        return false;
    }
    std::string err;
    return json_parse(text, out, err);
}

std::optional<LocalModel> find_local(const ModelRef & ref) {
    LocalModel m;

    if (ref.source == ModelRef::FILE) {
        std::error_code ec;
        if (!fs::exists(ref.file, ec)) {
            return std::nullopt;
        }
        m.model_path = ref.file;
        return m;
    }

    fs::path dir = ref.store_dir(models_root());
    if (!load_manifest(dir, m.manifest)) {
        return std::nullopt;
    }
    std::string model_file = m.manifest.get_str("model_file");
    if (model_file.empty()) {
        return std::nullopt;
    }
    m.dir = dir;
    m.model_path = dir / model_file;
    std::error_code ec;
    if (!fs::exists(m.model_path, ec)) {
        return std::nullopt;
    }
    std::string mmproj = m.manifest.get_str("mmproj_file");
    if (!mmproj.empty() && fs::exists(dir / mmproj, ec)) {
        m.mmproj_path = dir / mmproj;
    }
    return m;
}

void write_manifest(const fs::path & dir, const JValue & manifest) {
    fs::create_directories(dir);
    fs::path tmp = dir / "manifest.json.tmp";
    {
        std::ofstream f(tmp, std::ios::binary | std::ios::trunc);
        if (!f) {
            fail("cannot write " + tmp.string());
        }
        f << manifest.dump(2) << "\n";
    }
    fs::rename(tmp, dir / "manifest.json");
}

std::vector<ListedModel> list_models() {
    std::vector<ListedModel> out;
    fs::path root = models_root();
    std::error_code ec;
    if (!fs::exists(root, ec)) {
        return out;
    }

    for (auto it = fs::recursive_directory_iterator(root, ec);
         it != fs::recursive_directory_iterator(); it.increment(ec)) {
        if (ec) {
            break;
        }
        if (!it->is_regular_file(ec) || it->path().filename() != "manifest.json") {
            continue;
        }
        fs::path dir = it->path().parent_path();
        JValue man;
        if (!load_manifest(dir, man)) {
            continue;
        }
        ListedModel lm;
        lm.dir = dir;
        lm.name = man.get_str("name", dir.filename().string());
        lm.source = man.get_str("source", "?");
        lm.size = (uint64_t) man.get_num("size", 0);
        if (lm.size == 0) {
            for (auto & f : fs::directory_iterator(dir, ec)) {
                if (f.is_regular_file(ec) && f.path().extension() == ".gguf") {
                    lm.size += (uint64_t) f.file_size(ec);
                }
            }
        }
        lm.modified = man.get_str("pulled_at", "");
        out.push_back(std::move(lm));
    }

    std::sort(out.begin(), out.end(), [](const ListedModel & a, const ListedModel & b) {
        return a.name < b.name;
    });
    return out;
}

bool remove_model(const ModelRef & ref) {
    if (ref.source == ModelRef::FILE) {
        fail("refusing to delete a raw file path; remove it yourself if intended");
    }
    fs::path dir = ref.store_dir(models_root());
    std::error_code ec;
    if (!fs::exists(dir / "manifest.json", ec)) {
        return false;
    }
    fs::remove_all(dir, ec);
    if (ec) {
        fail("failed to remove " + dir.string() + ": " + ec.message());
    }
    // Tidy now-empty parents up to the models root.
    fs::path parent = dir.parent_path();
    while (parent != models_root() && fs::is_empty(parent, ec) && !ec) {
        fs::remove(parent, ec);
        parent = parent.parent_path();
    }
    return true;
}

} // namespace alpacca

File: alpacca/store.py
# Alpacca - local model store (~/.alpacca/models) and model references.
# MIT License. See LICENSE.
from __future__ import annotations

import contextlib
import json
import os
import re
import sys
import tempfile
import time
import unicodedata
from dataclasses import dataclass, field
from pathlib import Path


def alpacca_home() -> Path:
    env = os.environ.get("ALPACCA_HOME")
    if env:
        return Path(env)
    return Path.home() / ".alpacca"


def models_root() -> Path:
    return alpacca_home() / "models"


def _nicknames_file() -> Path:
    return alpacca_home() / "model-nicknames.json"


def _sanitize(part: str) -> str:
    out = re.sub(r"[^A-Za-z0-9._+-]", "_", part)
    return out if out not in ("", ".", "..") else "_"


@dataclass
class ModelRef:
    """Parsed model reference.

    "llama3.2:1b"           -> ollama registry, library/llama3.2:1b
    "ollama:user/name:tag"  -> ollama registry, user namespace
    "hf:org/repo:Q4_K_M"    -> Hugging Face repo + quant/file selector
    "org/repo"              -> Hugging Face
    "./model.gguf"          -> local file
    """
    source: str   # "ollama" | "hf" | "file"
    ns: str = ""
    name: str = ""
    tag: str = ""
    path: Path | None = None

    def display(self) -> str:
        if self.source == "file":
            return str(self.path)
        if self.source == "hf":
            s = f"hf:{self.ns}/{self.name}"
            return f"{s}:{self.tag}" if self.tag else s
        # must re-parse to the same ref: a bare "ns/name" is read as Hugging
        # Face, so a non-library namespace has to keep its disambiguator
        s = self.name if self.ns == "library" else f"ollama:{self.ns}/{self.name}"
        return f"{s}:{self.tag}" if self.tag != "latest" else s

    def store_dir(self) -> Path:
        root = models_root()
        if self.source == "hf":
            return root / "hf" / _sanitize(self.ns) / _sanitize(self.name) / \
                   _sanitize(self.tag or "default")
        if self.source == "ollama":
            return root / "ollama" / _sanitize(self.ns) / _sanitize(self.name) / \
                   _sanitize(self.tag)
        raise ValueError("file references have no store directory")


def parse_model_ref(raw: str) -> ModelRef:
    s = raw.strip()
    if not s:
        raise ValueError("empty model name")

    looks_path = (s.startswith(("/", "./", "../", "~/")) or
                  (os.name == "nt" and (re.match(r"^[A-Za-z]:", s) or
                                        s.startswith((".\\", "..\\", "\\\\")))) or
                  (s.lower().endswith(".gguf") and Path(s).expanduser().exists()))
    if looks_path:
        return ModelRef(source="file", path=Path(s).expanduser())

    forced_hf = False
    for p in ("hf:", "hf.co/", "huggingface.co/", "https://huggingface.co/"):
        if s.startswith(p):
            s = s[len(p):]
            forced_hf = True
            break
    forced_ollama = False
    if not forced_hf and s.startswith("ollama:"):
        s = s[len("ollama:"):]
        forced_ollama = True

    # split a trailing :tag (no '/' after the colon)
    name_part, tag = s, ""
    colon = s.rfind(":")
    if colon != -1 and "/" not in s[colon:]:
        name_part, tag = s[:colon], s[colon + 1:]

    parts = name_part.split("/")
    if any(not p.strip() for p in parts):
        raise ValueError(f"invalid model reference: '{raw}'")

    if forced_hf or (not forced_ollama and len(parts) >= 2):
        if len(parts) != 2:
            raise ValueError(f"Hugging Face references look like org/repo[:quant] (got '{raw}')")
        return ModelRef(source="hf", ns=parts[0], name=parts[1], tag=tag)

    if len(parts) == 1:
        return ModelRef(source="ollama", ns="library", name=parts[0], tag=tag or "latest")
    if len(parts) == 2:
        return ModelRef(source="ollama", ns=parts[0], name=parts[1], tag=tag or "latest")
    raise ValueError(f"Ollama references look like [user/]name[:tag] (got '{raw}')")


@dataclass
class LocalModel:
    model_path: Path
    dir: Path | None = None
    manifest: dict = field(default_factory=dict)


def find_local(ref: ModelRef) -> LocalModel | None:
    if ref.source == "file":
        assert ref.path is not None
        return LocalModel(model_path=ref.path) if ref.path.exists() else None
    d = ref.store_dir()
    mf = d / "manifest.json"
    if not mf.exists():
        return None
    try:
        manifest = json.loads(mf.read_text("utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    model_file = manifest.get("model_file", "")
    if not model_file or not (d / model_file).exists():
        return None
    return LocalModel(model_path=d / model_file, dir=d, manifest=manifest)


def write_manifest(d: Path, manifest: dict) -> None:
    d.mkdir(parents=True, exist_ok=True)
    tmp = d / "manifest.json.tmp"
    tmp.write_text(json.dumps(manifest, indent=2) + "\n", "utf-8")
    tmp.replace(d / "manifest.json")


def _clean_nickname(nickname: str) -> str:
    # nicknames are echoed to a terminal and written into JSON, so replace
    # anything invisible or direction-changing with a space before it can be
    # stored: Cc for ESC (which would emit ANSI escape sequences), Cf for the
    # bidi overrides that reorder the rest of the line and for zero-width
    # joiners, and Cs for lone surrogates arriving from argv
    text = "".join(" " if unicodedata.category(ch) in ("Cc", "Cf", "Cs") else ch
                   for ch in str(nickname))
    return " ".join(text.strip().split())


def _quarantine_nicknames(path: Path, err: Exception) -> None:
    """Move an unreadable nicknames file aside and say so, once."""
    spoiled = path.with_suffix(path.suffix + ".corrupt")
    try:
        os.replace(path, spoiled)
    except OSError:
        print(f"alpacca: warning: {path} is unreadable ({err}); "
              f"model nicknames are being ignored", file=sys.stderr)
        return
    print(f"alpacca: warning: {path} is unreadable ({err}); "
          f"moved it to {spoiled.name} and starting a new nickname file",
          file=sys.stderr)


def _read_nicknames() -> dict[str, str]:
    path = _nicknames_file()
    try:
        data = json.loads(path.read_text("utf-8"))
    except FileNotFoundError:
        return {}
    except OSError:
        return {}
    except (UnicodeDecodeError, json.JSONDecodeError, RecursionError) as e:
        # RecursionError is a RuntimeError, not a ValueError: deeply nested
        # JSON would otherwise fail every command that touches the map.
        # The file is readable but unusable - the next write would replace it,
        # so put it out of harm's way rather than destroying the only copy.
        _quarantine_nicknames(path, e)
        return {}
    if isinstance(data, dict) and isinstance(data.get("nicknames"), dict):
        data = data["nicknames"]
    if not isinstance(data, dict):
        return {}
    out: dict[str, str] = {}
    for nickname, target in data.items():
        nickname = _clean_nickname(str(nickname))
        target = str(target).strip()
        if not nickname or not target:
            continue
        try:
            out[nickname] = parse_model_ref(target).display()
        except ValueError:
            continue
    return out


def _write_nicknames(nicknames: dict[str, str]) -> None:
    path = _nicknames_file()
    if not nicknames:
        try:
            path.unlink()
        except FileNotFoundError:
            pass
        return
    path.parent.mkdir(parents=True, exist_ok=True)
    cleaned = {k: v for k, v in sorted(nicknames.items(), key=lambda kv: kv[0].lower())}
    payload = json.dumps({"nicknames": cleaned}, indent=2) + "\n"
    # a temp name shared between processes is worse than no temp file at all:
    # two writers interleave into it and replace() then publishes the garbage
    fd, tmp_name = tempfile.mkstemp(dir=str(path.parent),
                                    prefix=path.name + ".", suffix=".tmp")
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as fh:
            fh.write(payload)
            fh.flush()
            os.fsync(fh.fileno())
        os.replace(tmp_name, path)
    except BaseException:
        try:
            os.unlink(tmp_name)
        except OSError:
            pass
        raise


@contextlib.contextmanager
def _nicknames_lock():
    """Serialise the nickname read-modify-write across processes.

    Best-effort: an atomic write alone still loses updates, because two
    processes each read the old map and write back only their own entry.
    If locking is unavailable the mutation still proceeds - a lost alias is
    better than a command that refuses to run.
    """
    path = _nicknames_file()
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        fh = open(path.with_name(path.name + ".lock"), "a+b")
    except OSError:
        yield
        return
    locked = False
    try:
        try:
            if os.name == "nt":
                import msvcrt
                msvcrt.locking(fh.fileno(), msvcrt.LK_LOCK, 1)
            else:
                import fcntl
                fcntl.flock(fh.fileno(), fcntl.LOCK_EX)
            locked = True
        except (OSError, ImportError, ValueError):
            pass
        yield
    finally:
        if locked:
            try:
                if os.name == "nt":
                    import msvcrt
                    fh.seek(0)
                    msvcrt.locking(fh.fileno(), msvcrt.LK_UNLCK, 1)
                else:
                    import fcntl
                    fcntl.flock(fh.fileno(), fcntl.LOCK_UN)
            except (OSError, ImportError, ValueError):
                pass
        fh.close()


def list_nicknames() -> dict[str, str]:
    """Return nickname -> canonical model reference."""
    return dict(_read_nicknames())


def nickname_for_model(model_name: str) -> str:
    try:
        canonical = parse_model_ref(model_name).display()
    except ValueError:
        canonical = model_name.strip()
    for nickname, target in _read_nicknames().items():
        if target == canonical:
            return nickname
    return ""


def resolve_model_input(raw: str) -> str:
    """Resolve a CLI/menu model input to a canonical model reference.

    Installed canonical names win over nicknames so adding a nickname cannot
    shadow an existing model. If no installed model matches, an exact or
    unique case-insensitive nickname is accepted before falling back to the
    normal model-reference parser.

    File references are returned unchanged. Their display() form is a
    pathlib-normalised string ("./tiny" -> "tiny"), and callers re-parse what
    they get back - so returning display() here would silently reclassify a
    local path as a registry name.
    """
    s = raw.strip()
    if not s:
        raise ValueError("empty model name")
    ref: ModelRef | None
    try:
        ref = parse_model_ref(s)
    except ValueError:
        ref = None
    if ref is not None and ref.source == "file":
        return s
    if ref is not None and find_local(ref) is not None:
        return ref.display()

    cleaned = _clean_nickname(s)
    nicknames = _read_nicknames()
    if cleaned in nicknames:
        return nicknames[cleaned]
    matches = [target for nickname, target in nicknames.items()
               if nickname.lower() == cleaned.lower()]
    if len(set(matches)) == 1:
        return matches[0]
    if matches:
        raise ValueError(f"ambiguous model nickname: {raw}")
    if ref is not None:
        return ref.display()
    return parse_model_ref(s).display()


def set_model_nickname(model_name: str, nickname: str) -> tuple[str, str]:
    target = resolve_model_input(model_name)
    ref = parse_model_ref(target)
    if ref.source == "file" or find_local(ref) is None:
        raise ValueError(f"{target} is not an installed model")
    nickname = _clean_nickname(nickname)
    if not nickname:
        raise ValueError("empty model nickname")

    try:
        nick_ref = parse_model_ref(nickname)
    except ValueError:
        nick_ref = None  # not a parseable reference, so it cannot collide
    if nick_ref is not None and nick_ref.source == "file":
        raise ValueError(
            f"nickname '{nickname}' looks like a file path; "
            f"it could never be resolved back to a model")
    if nick_ref is not None and find_local(nick_ref) is not None \
            and nick_ref.display() != target:
        raise ValueError(
            f"nickname '{nickname}' conflicts with installed model {nick_ref.display()}")

    with _nicknames_lock():
        nicknames = _read_nicknames()
        for existing, existing_target in list(nicknames.items()):
            if existing_target == target:
                del nicknames[existing]
            elif existing.lower() == nickname.lower():
                raise ValueError(
                    f"nickname '{nickname}' already points to {existing_target}")
        nicknames[nickname] = target
        _write_nicknames(nicknames)
    return nickname, target


def clear_model_nickname(model_name: str) -> str:
    target = resolve_model_input(model_name)
    with _nicknames_lock():
        nicknames = _read_nicknames()
        removed = ""
        for nickname, existing_target in list(nicknames.items()):
            if existing_target == target:
                removed = nickname
                del nicknames[nickname]
        if removed:
            _write_nicknames(nicknames)
    return removed


def _remove_nicknames_for_model(model_name: str) -> None:
    with _nicknames_lock():
        nicknames = _read_nicknames()
        changed = False
        for nickname, target in list(nicknames.items()):
            if target == model_name:
                del nicknames[nickname]
                changed = True
        if changed:
            _write_nicknames(nicknames)


def now_iso8601() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


def human_size(n: float) -> str:
    for unit in ("B", "KB", "MB", "GB", "TB"):
        if n < 1024 or unit == "TB":
            return f"{n:.0f} {unit}" if unit == "B" else f"{n:.1f} {unit}"
        n /= 1024.0
    return f"{n:.1f} TB"


def list_models() -> list[dict]:
    out = []
    root = models_root()
    if not root.exists():
        return out
    nicknames = _read_nicknames()
    model_to_nickname = {target: nickname for nickname, target in nicknames.items()}
    for mf in sorted(root.rglob("manifest.json")):
        d = mf.parent
        try:
            manifest = json.loads(mf.read_text("utf-8"))
        except (OSError, json.JSONDecodeError):
            continue
        size = manifest.get("size", 0)
        if not size:
            size = sum(f.stat().st_size for f in d.glob("*.gguf"))
        name = manifest.get("name", d.name)
        # nickname_for_model canonicalises through parse_model_ref before
        # comparing; match on the same form so the two never disagree about
        # whether a model has an alias
        try:
            canonical = parse_model_ref(name).display()
        except ValueError:
            canonical = name
        out.append({
            "name": name,
            "nickname": (model_to_nickname.get(canonical)
                         or model_to_nickname.get(name, "")),
            "source": manifest.get("source", "?"),
            "size": int(size),
            "pulled_at": manifest.get("pulled_at", ""),
            "dir": d,
        })
    out.sort(key=lambda m: m["name"])
    return out


def remove_model(ref: ModelRef) -> bool:
    if ref.source == "file":
        raise ValueError("refusing to delete a raw file path; remove it yourself if intended")
    d = ref.store_dir()
    if not (d / "manifest.json").exists():
        return False
    name = ref.display()
    for f in sorted(d.rglob("*"), reverse=True):
        f.unlink() if f.is_file() else f.rmdir()
    d.rmdir()
    # the model is already gone, so nothing below here may report rm as failed:
    # not the alias bookkeeping, and not pruning the now-empty parents
    try:
        _remove_nicknames_for_model(name)
    except OSError:
        pass
    try:
        parent = d.parent
        while parent != models_root() and parent.exists() and not any(parent.iterdir()):
            parent.rmdir()
            parent = parent.parent
    except OSError:
        pass
    return True
